File: python/ch25/bandit_compare.py
# python/ch25/bandit_compare.py
"""AHC003 簡略グラフで辺選択バンディット 3 戦略を比較（UCB1 / Thompson / ε-greedy）."""
from __future__ import annotations

import heapq
import math
import random
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Tuple


def edge_key(u: int, v: int) -> Tuple[int, int]:
    return (u, v) if u < v else (v, u)


def load_sample(path: Path) -> Tuple[int, List[Tuple[int, int, float]], List[Tuple[int, int]]]:
    lines = path.read_text().splitlines()
    n, m = map(int, lines[0].split())
    edges: List[Tuple[int, int, float]] = []
    for i in range(1, m + 1):
        u, v, w = lines[i].split()
        edges.append((int(u), int(v), float(w)))
    q = int(lines[m + 1])
    queries = []
    for j in range(m + 2, m + 2 + q):
        s, t = map(int, lines[j].split())
        queries.append((s, t))
    return n, edges, queries


def build_adj(n: int, edges: List[Tuple[int, int, float]]) -> List[List[Tuple[int, float]]]:
    adj: List[List[Tuple[int, float]]] = [[] for _ in range(n)]
    for u, v, w in edges:
        adj[u].append((v, w))
        adj[v].append((u, w))
    return adj


def dijkstra(adj: List[List[Tuple[int, float]]], start: int, goal: int) -> Tuple[float, List[int]]:
    n = len(adj)
    dist = [math.inf] * n
    prev = [-1] * n
    dist[start] = 0.0
    pq: List[Tuple[float, int]] = [(0.0, start)]
    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        if u == goal:
            break
        for v, w in adj[u]:
            nd = d + w
            if nd < dist[v]:
                dist[v] = nd
                prev[v] = u
                heapq.heappush(pq, (nd, v))
    if dist[goal] == math.inf:
        return math.inf, []
    path: List[int] = []
    cur = goal
    while cur != -1:
        path.append(cur)
        cur = prev[cur]
    path.reverse()
    return dist[goal], path


@dataclass
class EdgeStats:
    pulls: int = 0
    reward_sum: float = 0.0
    alpha: float = 1.0
    beta: float = 1.0


class BanditSolver:
    def __init__(
        self,
        n: int,
        topo: List[Tuple[int, int]],
        strategy: str,
        rng: random.Random,
        epsilon: float = 0.15,
    ) -> None:
        self.n = n
        self.topo = topo
        self.strategy = strategy
        self.rng = rng
        self.epsilon = epsilon
        self.est: Dict[Tuple[int, int], float] = {edge_key(u, v): 1.0 for u, v in topo}
        self.stats: Dict[Tuple[int, int], EdgeStats] = {
            k: EdgeStats() for k in self.est
        }
        self.total_pulls = 0

    def adj(self) -> List[List[Tuple[int, float]]]:
        adj: List[List[Tuple[int, float]]] = [[] for _ in range(self.n)]
        for (u, v), w in self.est.items():
            bonus = self.exploration_bonus((u, v))
            adj[u].append((v, max(0.01, w - bonus)))
            adj[v].append((u, max(0.01, w - bonus)))
        return adj

    def exploration_bonus(self, key: Tuple[int, int]) -> float:
        st = self.stats[key]
        if st.pulls == 0:
            return 2.0
        if self.strategy == "ucb1":
            return 0.5 * math.sqrt(math.log(self.total_pulls + 1) / st.pulls)
        if self.strategy == "thompson":
            sample = self.rng.betavariate(st.alpha, st.beta)
            mean = st.reward_sum / st.pulls if st.pulls else 0.5
            return 0.8 * (sample - mean)
        if self.strategy == "epsilon_greedy":
            if self.rng.random() < self.epsilon:
                return 1.5
            return 0.0
        return 0.0

    def pick_focus_edge(self, path: List[int]) -> Tuple[int, int]:
        edges_on_path = [edge_key(path[i], path[i + 1]) for i in range(len(path) - 1)]
        if not edges_on_path:
            return next(iter(self.est))
        if self.strategy == "ucb1":
            return max(
                edges_on_path,
                key=lambda k: self.stats[k].reward_sum / max(1, self.stats[k].pulls)
                + math.sqrt(2 * math.log(self.total_pulls + 1) / max(1, self.stats[k].pulls)),
            )
        if self.strategy == "thompson":
            return max(
                edges_on_path,
                key=lambda k: self.rng.betavariate(self.stats[k].alpha, self.stats[k].beta),
            )
        if self.rng.random() < self.epsilon:
            return self.rng.choice(edges_on_path)
        return max(edges_on_path, key=lambda k: self.stats[k].reward_sum / max(1, self.stats[k].pulls))

    def update_estimates(self, path: List[int], predicted: float, actual: float) -> None:
        if len(path) < 2 or predicted <= 0:
            return
        ratio = actual / predicted
        for i in range(len(path) - 1):
            k = edge_key(path[i], path[i + 1])
            self.est[k] = max(0.1, self.est.get(k, 1.0) * ratio)
        focus = self.pick_focus_edge(path)
        err = max(0.0, min(1.0, abs(actual - predicted) / max(actual, 1.0)))
        reward = 1.0 - err
        st = self.stats[focus]
        st.pulls += 1
        st.reward_sum += reward
        st.alpha += reward
        st.beta += 1.0 - reward
        self.total_pulls += 1
        self.est[focus] = max(0.1, self.est[focus] * (0.9 + 0.2 * reward))

    def answer(self, s: int, t: int) -> Tuple[float, List[int]]:
        return dijkstra(self.adj(), s, t)

    def run_offline(
        self,
        true_adj: List[List[Tuple[int, float]]],
        queries: List[Tuple[int, int]],
    ) -> Tuple[float, float]:
        total_pred = 0.0
        total_true = 0.0
        for s, t in queries:
            pred, path = self.answer(s, t)
            true_len, true_path = dijkstra(true_adj, s, t)
            self.update_estimates(true_path, pred, true_len)
            total_pred += pred
            total_true += true_len
        return total_pred, total_true


def main() -> None:
    root = Path(__file__).resolve().parents[2]
    path = Path(sys.argv[1]) if len(sys.argv) > 1 else root / "data" / "ch25" / "sample.txt"
    n, edges, queries = load_sample(path)
    true_adj = build_adj(n, edges)
    topo = [(u, v) for u, v, _ in edges]
    rng = random.Random(42)

    print(f"graph n={n} m={len(edges)} queries={len(queries)}", file=sys.stderr)
    results = []
    for name in ("ucb1", "thompson", "epsilon_greedy"):
        solver = BanditSolver(n, topo, name, rng)
        pred, true = solver.run_offline(true_adj, queries)
        gap = pred - true
        results.append((name, pred, true, gap))
        print(
            f"strategy={name} sum_pred={pred:.4f} sum_true={true:.4f} gap={gap:.4f}",
            file=sys.stderr,
        )

    best = min(results, key=lambda x: x[3])
    print(f"best_gap={best[0]} (smallest pred-true gap)", file=sys.stderr)


if __name__ == "__main__":
    main()

File: go/ch25/bandit_compare.go
// go/ch25/bandit_compare.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

type edge struct {
	u, v int
	w    float64
}

type EdgeStats struct {
	pulls     int
	rewardSum float64
	alpha     float64
	beta      float64
}

type BanditSolver struct {
	n          int
	topo       [][2]int
	strategy   string
	rng        *rand.Rand
	epsilon    float64
	est        map[[2]int]float64
	stats      map[[2]int]*EdgeStats
	totalPulls int
}

func edgeKey(u, v int) [2]int {
	if u < v {
		return [2]int{u, v}
	}
	return [2]int{v, u}
}

func loadSample(path string) (int, []edge, [][2]int, error) {
	f, err := os.Open(path)
	if err != nil {
		return 0, nil, nil, err
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	var lines []string
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}
	if err := sc.Err(); err != nil {
		return 0, nil, nil, err
	}
	p := strings.Fields(lines[0])
	n, _ := strconv.Atoi(p[0])
	m, _ := strconv.Atoi(p[1])
	edges := make([]edge, m)
	topo := make([][2]int, m)
	for i := 0; i < m; i++ {
		q := strings.Fields(lines[1+i])
		u, _ := strconv.Atoi(q[0])
		v, _ := strconv.Atoi(q[1])
		w, _ := strconv.ParseFloat(q[2], 64)
		edges[i] = edge{u, v, w}
		topo[i] = [2]int{u, v}
	}
	q, _ := strconv.Atoi(lines[1+m])
	queries := make([][2]int, q)
	for j := 0; j < q; j++ {
		qp := strings.Fields(lines[2+m+j])
		s, _ := strconv.Atoi(qp[0])
		t, _ := strconv.Atoi(qp[1])
		queries[j] = [2]int{s, t}
	}
	return n, edges, queries, nil
}

func buildAdj(n int, edges []edge) [][]struct {
	to int
	w  float64
} {
	adj := make([][]struct {
		to int
		w  float64
	}, n)
	for _, e := range edges {
		adj[e.u] = append(adj[e.u], struct {
			to int
			w  float64
		}{e.v, e.w})
		adj[e.v] = append(adj[e.v], struct {
			to int
			w  float64
		}{e.u, e.w})
	}
	return adj
}

func dijkstra(adj [][]struct {
	to int
	w  float64
}, start, goal int) (float64, []int) {
	n := len(adj)
	dist := make([]float64, n)
	prev := make([]int, n)
	used := make([]bool, n)
	for i := range dist {
		dist[i] = math.Inf(1)
		prev[i] = -1
	}
	dist[start] = 0
	for iter := 0; iter < n; iter++ {
		u := -1
		best := math.Inf(1)
		for i := 0; i < n; i++ {
			if !used[i] && dist[i] < best {
				best = dist[i]
				u = i
			}
		}
		if u == -1 {
			break
		}
		used[u] = true
		if u == goal {
			break
		}
		for _, e := range adj[u] {
			nd := dist[u] + e.w
			if nd < dist[e.to] {
				dist[e.to] = nd
				prev[e.to] = u
			}
		}
	}
	if dist[goal] == math.Inf(1) {
		return math.Inf(1), nil
	}
	path := []int{}
	cur := goal
	for cur != -1 {
		path = append([]int{cur}, path...)
		cur = prev[cur]
	}
	return dist[goal], path
}

func newBanditSolver(n int, topo [][2]int, strategy string, rng *rand.Rand) *BanditSolver {
	est := make(map[[2]int]float64)
	stats := make(map[[2]int]*EdgeStats)
	for _, e := range topo {
		k := edgeKey(e[0], e[1])
		est[k] = 1.0
		stats[k] = &EdgeStats{alpha: 1, beta: 1}
	}
	return &BanditSolver{
		n: n, topo: topo, strategy: strategy, rng: rng, epsilon: 0.15,
		est: est, stats: stats,
	}
}

func (s *BanditSolver) explorationBonus(key [2]int) float64 {
	st := s.stats[key]
	if st.pulls == 0 {
		return 2.0
	}
	switch s.strategy {
	case "ucb1":
		return 0.5 * math.Sqrt(math.Log(float64(s.totalPulls+1))/float64(st.pulls))
	case "thompson":
		sample := betaSample(s.rng, st.alpha, st.beta)
		mean := st.rewardSum / float64(st.pulls)
		return 0.8 * (sample - mean)
	case "epsilon_greedy":
		if s.rng.Float64() < s.epsilon {
			return 1.5
		}
	}
	return 0
}

func betaSample(rng *rand.Rand, alpha, beta float64) float64 {
	// 簡易: 平均 + 小ノイズ（本番は math/rand 以外でも可）
	return alpha/(alpha+beta) + (rng.Float64()-0.5)*0.2
}

func (s *BanditSolver) adj() [][]struct {
	to int
	w  float64
} {
	adj := make([][]struct {
		to int
		w  float64
	}, s.n)
	for k, w := range s.est {
		bonus := s.explorationBonus(k)
		adjW := w - bonus
		if adjW < 0.01 {
			adjW = 0.01
		}
		adj[k[0]] = append(adj[k[0]], struct {
			to int
			w  float64
		}{k[1], adjW})
		adj[k[1]] = append(adj[k[1]], struct {
			to int
			w  float64
		}{k[0], adjW})
	}
	return adj
}

func (s *BanditSolver) pickFocus(path []int) [2]int {
	var edges [][2]int
	for i := 0; i+1 < len(path); i++ {
		edges = append(edges, edgeKey(path[i], path[i+1]))
	}
	if len(edges) == 0 {
		for k := range s.est {
			return k
		}
	}
	bestK := edges[0]
	bestScore := -1.0
	for _, k := range edges {
		st := s.stats[k]
		score := 0.0
		switch s.strategy {
		case "ucb1":
			score = st.rewardSum/float64(max(1, st.pulls)) +
				math.Sqrt(2*math.Log(float64(s.totalPulls+1))/float64(max(1, st.pulls)))
		case "thompson":
			score = betaSample(s.rng, st.alpha, st.beta)
		default:
			score = st.rewardSum / float64(max(1, st.pulls))
			if s.rng.Float64() < s.epsilon {
				return k
			}
		}
		if score > bestScore {
			bestScore = score
			bestK = k
		}
	}
	return bestK
}

func (s *BanditSolver) updateEstimates(path []int, predicted, actual float64) {
	if len(path) < 2 || predicted <= 0 {
		return
	}
	ratio := actual / predicted
	for i := 0; i+1 < len(path); i++ {
		k := edgeKey(path[i], path[i+1])
		s.est[k] = math.Max(0.1, s.est[k]*ratio)
	}
	focus := s.pickFocus(path)
	err := math.Abs(actual-predicted) / math.Max(actual, 1)
	if err < 0 {
		err = 0
	}
	if err > 1 {
		err = 1
	}
	reward := 1.0 - err
	st := s.stats[focus]
	st.pulls++
	st.rewardSum += reward
	st.alpha += reward
	st.beta += 1.0 - reward
	s.totalPulls++
	s.est[focus] = math.Max(0.1, s.est[focus]*(0.9+0.2*reward))
}

func (s *BanditSolver) runOffline(trueAdj [][]struct {
	to int
	w  float64
}, queries [][2]int) (float64, float64) {
	totalPred, totalTrue := 0.0, 0.0
	for _, q := range queries {
		pred, _ := dijkstra(s.adj(), q[0], q[1])
		trueLen, truePath := dijkstra(trueAdj, q[0], q[1])
		s.updateEstimates(truePath, pred, trueLen)
		totalPred += pred
		totalTrue += trueLen
	}
	return totalPred, totalTrue
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	path := "data/ch25/sample.txt"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	n, edges, queries, err := loadSample(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	trueAdj := buildAdj(n, edges)
	topo := make([][2]int, len(edges))
	for i, e := range edges {
		topo[i] = [2]int{e.u, e.v}
	}
	rng := rand.New(rand.NewSource(42))
	fmt.Fprintf(os.Stderr, "graph n=%d m=%d queries=%d\n", n, len(edges), len(queries))

	bestName := ""
	bestGap := math.MaxFloat64
	for _, name := range []string{"ucb1", "thompson", "epsilon_greedy"} {
		solver := newBanditSolver(n, topo, name, rng)
		pred, true := solver.runOffline(trueAdj, queries)
		gap := pred - true
		fmt.Fprintf(os.Stderr, "strategy=%s sum_pred=%.4f sum_true=%.4f gap=%.4f\n",
			name, pred, true, gap)
		if gap < bestGap {
			bestGap = gap
			bestName = name
		}
	}
	fmt.Fprintf(os.Stderr, "best_gap=%s (smallest pred-true gap)\n", bestName)
}
